include/syscall.h

```diff
@@ -57,59 +106,141 @@
+ * @return nonzero if the CPU is currently running with user permissions
  */
-#define _SYSCALL_VERIFY(expr, ssf) \
-	do { \
-		if (!(expr)) { \
-			printk("FATAL: syscall failed check: " #expr "\n"); \
-			_arch_syscall_oops(ssf); \
-		} \
-	} while (0)
+static inline int _arch_is_user_context(void);
 
 /**
- * @brief Runtime check that a pointer is a kernel object of expected type
+ * Indicate whether the CPU is currently in user mode
  *
- * Passes along arguments to _k_object_validate() and triggers a kernel oops
- * if the object wasn't valid or had incorrect permissions.
- *
- * @param ptr Untrusted kernel object pointer
- * @param type Expected kernel object type
- * @param init Whether this is an init function handler
- * @param ssf Syscall stack frame argument passed to the handler function
+ * @return nonzero if the CPU is currently running with user permissions
  */
-#define _SYSCALL_IS_OBJ(ptr, type, init, ssf) \
-	_SYSCALL_VERIFY(!_k_object_validate((void *)ptr, type, init), ssf)
-
-/* Convenience macros for handler implementations */
-#define _SYSCALL_ARG0	ARG_UNUSED(arg1); ARG_UNUSED(arg2); ARG_UNUSED(arg3); \
-			ARG_UNUSED(arg4); ARH_UNUSED(arg5); ARG_UNUSED(arg6)
+static inline int _is_user_context(void)
+{
+	return _arch_is_user_context();
+}
 
-#define _SYSCALL_ARG1	ARG_UNUSED(arg2); ARG_UNUSED(arg3); ARG_UNUSED(arg4); \
-			ARG_UNUSED(arg5); ARG_UNUSED(arg6)
-
-#define _SYSCALL_ARG2	ARG_UNUSED(arg3); ARG_UNUSED(arg4); ARG_UNUSED(arg5); \
-			ARG_UNUSED(arg6)
-
-#define _SYSCALL_ARG3	ARG_UNUSED(arg4); ARG_UNUSED(arg5); ARG_UNUSED(arg6)
+/*
+ * Helper data structures for system calls with large argument lists
+ */
 
+struct _syscall_7_args {
+	u32_t arg6;
+	u32_t arg7;
+};
+
+struct _syscall_8_args {
+	u32_t arg6;
+	u32_t arg7;
+	u32_t arg8;
+};
+
+struct _syscall_9_args {
+	u32_t arg6;
+	u32_t arg7;
+	u32_t arg8;
+	u32_t arg9;
+};
+
+struct _syscall_10_args {
+	u32_t arg6;
+	u32_t arg7;
+	u32_t arg8;
+	u32_t arg9;
+	u32_t arg10;
+};
 
-#define _SYSCALL_ARG4	ARG_UNUSED(arg5); ARG_UNUSED(arg6)
+/*
+ * Interfaces for invoking system calls
+ */
 
-#define _SYSCALL_ARG5	ARG_UNUSED(arg6)
+static inline u32_t _arch_syscall_invoke0(u32_t call_id);
+
+static inline u32_t _arch_syscall_invoke1(u32_t arg1, u32_t call_id);
+
+static inline u32_t _arch_syscall_invoke2(u32_t arg1, u32_t arg2,
+					  u32_t call_id);
+
+static inline u32_t _arch_syscall_invoke3(u32_t arg1, u32_t arg2, u32_t arg3,
+					  u32_t call_id);
+
+static inline u32_t _arch_syscall_invoke4(u32_t arg1, u32_t arg2, u32_t arg3,
+					  u32_t arg4, u32_t call_id);
+
+static inline u32_t _arch_syscall_invoke5(u32_t arg1, u32_t arg2, u32_t arg3,
+					  u32_t arg4, u32_t arg5,
+					  u32_t call_id);
+
+static inline u32_t _arch_syscall_invoke6(u32_t arg1, u32_t arg2, u32_t arg3,
+					  u32_t arg4, u32_t arg5, u32_t arg6,
+					  u32_t call_id);
+
+static inline void _syscall_invoke7(u32_t arg1, u32_t arg2, u32_t arg3,
```

Comment:
Any reasons why these syscall invocation functions for large number of parameters are void functions?

Reply:
whoops, good catch